### PR TITLE
Updated configgrpc ToDialOptions and confighttp ToClient apis to take extensions configuration map 

### PR DESCRIPTION
**Pull request by pavankrish123**
This PR is a port of the configfrpc's ToDialOptions() and confighttp ToClient() from PR #3128  in a piece meal fashion. 

The following are the changes
- Refactored configgrpc.PerRPCAuth as extension implementing configauth.GrpcClientAuthenticator
- Plugged in extensions configuration to all the grpc based clients in the core (OTLP, OpenCensus, Jaeger, JaegerReceiver)
- Plugged in extensions configuration to all the HTTP based clients in the core (Zipkin, OTLPHTTP)

Link to tracking Issue:
 #3282 #3276

Testing:
Unit tests, [manual test described (for only oidc)](#3128 (comment))

### CODE REVIEW

**Comment by pavankrish123** [comment]
cc: @bogdandrutu @jpkrohling @gramidt .

These changes will break the contrib and need updates to clients on contrib side in the similar fashion.

File: config/configgrpc/configgrpc.go
Comment (jpkrohling) [inline config/configgrpc/configgrpc.go:97]: Food for thought: can't we keep this for a couple of releases, deprecating it first, and then completely removing it? I think it should be possible to build an Auth struct out of the information from the PerRPCAuth instance.

File: config/confighttp/confighttp.go
Comment (jpkrohling) [inline config/confighttp/confighttp.go:120]:
```suggestion
// Custom RoundTripper that adds headers.
```

File: extension/bearertokenauthextension/README.md
Comment (jpkrohling) [inline extension/bearertokenauthextension/README.md:3]:
```suggestion
This extension implements `configauth.GRPCClientAuthenticator` and is to be used in gRPC receivers inside the `auth` settings as a means
```

File: extension/bearertokenauthextension/README.md
Comment (jpkrohling) [inline extension/bearertokenauthextension/README.md:4]:
```suggestion
to embed a static token for every RPC call that will be made.
```

File: extension/bearertokenauthextension/README.md
Comment (jpkrohling) [inline extension/bearertokenauthextension/README.md:14]:
```suggestion
  RPC metadata.
```

**Review by jpkrohling — COMMENTED** [review]
This is mainly the code that has been reviewed before, right? It looks good to me, just found a couple of minor issues.

I left a comment about whether we'd want to keep backward compatibility for the PerRPCAuth option. I'll let you decide on that, and it's OK to break compatibility at this phase. In fact, it's commonly done by other components. The argument I have *for* doing this change in a backward compatible manner is to build the mindset that will soon be required for such changes.

**Comment by jpkrohling** [comment]
I think this should be merged only when we have a PR ready to change the contrib components that will be broken by this. That other PR can then be updated to reference the latest core once this PR here gets merged.

**Comment by gramidt** [comment]
@jpkrohling @pavankrish123 - I'll have a PR ready to be rebased when we're ready. This should be straightforward given that all related PRs of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3507 have been merged.

@jpkrohling - While I agree with the backwards compatible mindset for the 1.0 release, I do not believe it's fair that we ask for this after we all approved the prior PRs and previously merged these changes.

**Comment by pavankrish123** [comment]
Thanks @gramidt, I will keep the PR as is. Let me know when we are done from contrib. I have updated the README suggestions from @jpkrohling.

**Comment by gramidt** [comment]
That's excellent news, @pavankrish123!

@jpkrohling @pavankrish123 - I have started the the changes needed for contrib in #3694; however, I'll need to wait until this is merged to build, run tests, and fix any syntax errors I had while adding method calls that weren't valid without these changes.

File: config/configgrpc/configgrpc.go
Comment (pavankrish123) [reply config/configgrpc/configgrpc.go:97]: Keeping the PR as is per @gramidt suggestion below. @jpkrohling please let me know, I can always open another PR with PerRPCAuth otherwise. I do think I think it is ok to break the compatibility for Auth as we move forward towards 1.0 and enforce backward compatibility afterwards.

**Comment by pavankrish123** [comment]
@tigrannajaryan @bogdandrutu let me know if you need anything else.

**Review by gramidt — APPROVED** [review]

**Comment by gramidt** [comment]
> That's excellent news, @pavankrish123!
> 
> @jpkrohling @pavankrish123 - I have started the the changes needed for contrib in [open-telemetry/opentelemetry-collector-contrib#3694](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3694); however, I'll need to wait until this is merged to build, run tests, and fix any syntax errors I had while adding method calls that weren't valid without these changes.

cc @bogdandrutu @tigrannajaryan

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Re-running tests, as the load test failed due to a known intermittent test failure.

**Comment by tigrannajaryan** [comment]
@pavankrish123 please rebase from main. I want to make sure the only failures on contrib-tests job are coming from the changes in this PR.

@gramidt to confirm https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3694 is lined up to be merged immediately once this PR is merged, right? You will only need to update the core version to latest, correct?

**Comment by tigrannajaryan** [comment]
> Re-running tests, as the load test failed due to a known intermittent test failure.

@jpkrohling Can you please record a bug or if it needs an increase in resource limits make that increase?

**Comment by jpkrohling** [comment]
There's one issue already, and I'll send a PR soon to increase the limits.

https://github.com/open-telemetry/opentelemetry-collector/issues/2911

**Review by tigrannajaryan — APPROVED** [review]
LGTM.
@bogdandrutu any concerns? I believe this moves in the desired smaller increment now and is ready to be merged.

**Comment by gramidt** [comment]
> @pavankrish123 please rebase from main. I want to make sure the only failures on contrib-tests job are coming from the changes in this PR.
> 
> 
> 
> @gramidt to confirm https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3694 is lined up to be merged immediately once this PR is merged, right? You will only need to update the core version to latest, correct?

That is the plan; however, i will not know until I'm able to update core, since I added code that is not valid until this is merged and core is updated. If there's other core changes that require additional work, I can not speak to that effort. I will also be out of the office for another 3 hrs, but can work on it immediately once I return.

**Comment by jpkrohling** [comment]
@gramidt there's one way to do it (examples using `gh` CLI tool)

1. `cd github.com/open-telemetry/opentelemetry-collector`
1. `$ gh pr checkout 3340` 
1. `cd github.com/open-telemetry/opentelemetry-collector-contrib`
1. `$ gh pr checkout 3694`
1. `for file in $(find . -name go.mod); do go mod edit -replace go.opentelemetry.io/collector@v0.27.1-0.20210602134904-017aa1dad750=/home/jpkroehling/Projects/src/github.com/open-telemetry/opentelemetry-collector $file ; done` (replace the local path on this last one)

I just did that and `make test` passed. `make all` also completed without errors.

**Comment by jpkrohling** [comment]
With your permission, @gramidt, I'm updating this PR with the rebase I did locally to test the commands I mentioned earlier.

**Comment by jpkrohling** [comment]
Actually, looks like I can't update this PR, as I'm not a maintainer :-) Rebasing to the latest main caused only one conflict, which should be easy to solve.

**Comment by gramidt** [comment]
Thanks @jpkrohling! I'll be back home sooner than originally planned, so I should be able to complete it relatively soon. Also, I can't believe I spaced out and forgot about using my local core collector. Haha

**Comment by pavankrish123** [comment]
Updating now.

**Comment by tigrannajaryan** [comment]
This is good to go, merging.

**Comment by pavankrish123** [comment]
Thanks @tigrannajaryan - @gramidt we are all set to commit on contrib

**Comment by tigrannajaryan** [comment]
Thanks @pavankrish123 !

**Pull request merged**
